cmd/multifile.go

```diff
@@ -0,0 +1,182 @@
+	MultiFileCMD.PersistentFlags().BoolVar(&MultifileSerializePerHostDownloads, optname.SerializePerHostDownloads, true, "Whether to serialize downloads from the same host")
+	err := viper.BindPFlags(MultiFileCMD.PersistentFlags())
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+}
+
+type manifestEntry struct {
+	url  string
+	dest string
+}
+
+func execMultifile(cmd *cobra.Command, args []string) error {
+	var reader *bufio.Reader
+	// if manifest file is '-', read from stdin
+	manifestPath := args[0]
+	if manifestPath == "-" {
+		reader = bufio.NewReader(os.Stdin)
+	} else {
+		// check that the manifest file exists
+		_, err := os.Stat(manifestPath)
+		if os.IsNotExist(err) {
+			return fmt.Errorf("manifest file %s does not exist", manifestPath)
+		}
+		manifestFile, err := os.Open(manifestPath)
+		if err != nil {
+			return fmt.Errorf("error opening manifest file %s: %w", manifestPath, err)
+		}
+		defer manifestFile.Close()
+		reader = bufio.NewReader(manifestFile)
+	}
+	// process the manifest file into a map of hosts to url/file pairs with processManifest
+	buffer := make([]string, 0)
+	for {
+		readString, err := reader.ReadString('\n')
+		// break on EOF or empty line
+		if err != nil || readString == "\n" {
+			break
+		}
+		buffer = append(buffer, readString)
+	}
+	manifest, err := processManifest(buffer)
+	if err != nil {
+		return fmt.Errorf("error processing manifest file %s: %w", manifestPath, err)
+	}
+	// download each host's files in parallel
+	var eg errgroup.Group
+	for host, entries := range manifest {
+		err := downloadFilesFromHost(&eg, entries)
+		if err != nil {
+			return fmt.Errorf("error initiating download of files from host %s: %w", host, err)
+		}
+	}
+	err = eg.Wait()
+	if err != nil {
+		return fmt.Errorf("error downloading files: %w", err)
+	}
+	return nil
+}
+
+func processManifest(buffer []string) (map[string][]manifestEntry, error) {
+	// track the urls and dests for each host, do not allow for duplicate urls with different dests
+	seenURLs := make(map[string]string)
+	manifestMap := make(map[string][]manifestEntry)
+	// read the manifest file line by line
+	for _, line := range buffer {
+
+		// split the line into url and dest
+		var urlString, dest string
+		_, err := fmt.Sscanf(line, "%s %s", &urlString, &dest)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing manifest invalid line format %s: %w", line, err)
+		}
+		// check URL is not in seenURLs
+		if seenDest, ok := seenURLs[urlString]; ok {
+			if seenDest != dest {
+				return nil, fmt.Errorf("duplicate url %s with different dests: %s and %s", urlString, seenDest, dest)
+			}
+		}
+		// add the url to seenURLs
+		seenURLs[urlString] = dest
+		// get the host from the url
+		parsedURL, err := url.Parse(urlString)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing url %s: %w", urlString, err)
+		}
+		host := parsedURL.Host
+		// add the url/dest pair to the manifestMap
+		if entries, ok := manifestMap[host]; !ok {
```

Comment:
we should include the scheme too, because requests to http://example.com and https://example.com will use different connections

Reply:
The only caveat is that this will not allow for the break on an empty line with a newline. I think you're right though, let me fix this to readfile.

Reply:
Good callout.

Reply:
ResolveReference takes a URL input. I think we should use a sprintf for url.Scheme and url.Host